### ISSUE TITLE
LineLength gets triggered on multi-line strings for Google Checks

### ISSUE THREAD

**Issue by ghost** [checkstyle/checkstyle#11867]
We have a sample of CSV input that we used for documenting REST APIs via OpenAPI, using the shiny new Java 17 multi-line strings
(instead of the ugly multi-line manual strings we had before)

```
  static final String LOCATION_CSV_SAMPLE =
      """
          locationId,label,regionId,regionLabel,vendorId,vendorLabel,address,address2,city,stateProvinceCode,zipCode,countryCode,latitude,longitude
          ST001,Station 001,ZONE1,Zone 1,CP1,Competitor 1,123 Street,Unit 2,Houston,TX,77033,US,29.761496813335178,-95.53049214204984
          ST002,Station 002,ZONE2,,CP2,,668 Street,Unit 23,San Jose,CA,95191,US,37.35102477242508,-121.9209934020318
          """;
```

the LineLength check gets triggerd on this and blocks us from using the benefits of multi-line strings like this.

I think this check should be more flexible and recognize that multi-line strings like this could be used for documentation, etc. and not get triggered. 

Otherwise it may nullify one of the greatest benefits of the new multi-line Java 17 strings.

```
public class LineLength {

  /** somejavadoc. */
  public static void main(String[] args) {
    
    // SHOULD NOT VIOLATE, THIS IS TEXT-BLOCK
    String textBlock = """ 
      This STRing will exceed the column limit of 100 but checktyle will not show many error because it is allowed 
      Random String
      google config allows text block to exceed limit of 100 line length. it improves readability and stuff.
      same is true for this line also. same is true for this line as well. same is true for this line too. 
      one more Random String.
        """;
     
    // OK to give violation as this is normal string
    String normalString = "a very long string is heavy to parse and takes so much processing power. it is not ideal to make long strings";
  }
}

$ java -jar checkstyle-10.25.0-all.jar -c /google_checks.xml LineLength.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:1:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:7: Line is longer than 100 characters (found 115). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:9: Line is longer than 100 characters (found 108). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:10: Line is longer than 100 characters (found 107). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:14: Line is longer than 100 characters (found 138). [LineLength]
Audit done.
```

Google style is updated to mention this
https://github.com/checkstyle/checkstyle/issues/11867#issuecomment-2954057044

**Comment by strkkk** [comment]
>  check should be more flexible and recognize that multi-line strings like this could be used for documentation, etc. and not get triggered.

@jfurmankiewiczpros LineLength is not aware of java syntax, therefore it cannot understand context of the line. And it is completely impossible to "teach" it whether line is a documentation or not. Basically, it is just a text.
> blocks us from using the benefits of multi-line strings like this.

I recommend you to just suppress such violations.

**Comment by nick-mancuso** [comment]
A good way to suppress such violations would be to use `SuppressWithPlainTextCommentFilter`:

```
➜  java git:(master) ✗ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="fileExtensions" value="java,js,css,xml"/>
    <module name="LineLength">
        <property name="max" value="100"/>
    </module>
    <module name="SuppressWithPlainTextCommentFilter">
        <property name="offCommentFormat" value='"""'/>
        <property name="onCommentFormat" value='""";'/>
    </module>
    <module name="TreeWalker">
    </module>
</module>

➜  java git:(master) ✗ cat Test.java 
public class Test {
    static final String LOCATION_CSV_SAMPLE =
            """
                locationId,label,regionId,regionLabel,vendorId,vendorLabel,address,address2,city,stateProvinceCode,zipCode,countryCode,latitude,longitude
                ST001,Station 001,ZONE1,Zone 1,CP1,Competitor 1,123 Street,Unit 2,Houston,TX,77033,US,29.761496813335178,-95.53049214204984
                ST002,Station 002,ZONE2,,CP2,,668 Street,Unit 23,San Jose,CA,95191,US,37.35102477242508,-121.9209934020318
                """;
}

➜  java git:(master) ✗ java -jar checkstyle-10.3.1-all.jar -c config.xml Test.java
Starting audit...
Audit done.


```

**Comment by nick-mancuso** [comment]
With slightly different formatting for your text block declaration, you can use `SuppressWithPlainTextCommentFilter` to suppress these warnings:

```
➜  java git:(master) ✗ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="fileExtensions" value="java,js,css,xml"/>
    <module name="LineLength">
        <property name="max" value="100"/>
    </module>
    <module name="SuppressWithPlainTextCommentFilter">
        <property name="offCommentFormat" value='=\s+"""'/>
        <property name="onCommentFormat" value='^\s+""";'/>
    </module>
    <module name="TreeWalker">
    </module>
</module>

➜  java git:(master) ✗ cat Test.java 
public class Test {
    static final String LOCATION_CSV_SAMPLE = """
                locationId,label,regionId,regionLabel,vendorId,vendorLabel,address,address2,city,stateProvinceCode,zipCode,countryCode,latitude,longitude
                ST001,Station 001,ZONE1,Zone 1,CP1,Competitor 1,123 Street,Unit 2,Houston,TX,77033,US,29.761496813335178,-95.53049214204984
                ST002,Station 002,ZONE2,,CP2,,668 Street,Unit 23,San Jose,CA,95191,US,37.35102477242508,-121.9209934020318
                """;

    final int asdofiaweifnowaeinaluadfliusabdflabdslivubaluvfbldaiubvfuabdvfualvubalduvfbladubvfladbvf = 4;

}%                                                                                                              

 ➜  java git:(master) ✗ java -jar checkstyle-10.3.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/Tester/src/main/java/Test.java:8: Line is longer than 100 characters (found 107). [LineLength]
Audit done.
Checkstyle ends with 1 errors.
➜

```

**Comment by mohitsatr** [comment]
> The contents of a text block may exceed the column limit

we should the add the suppression to the google config. google-style allows text to exceed  the column limit.

**Comment by mohitsatr** [comment]
@Zopsss @romani 
I tried the following suppression, but got exception
```
  </module>
      <module name="SuppressionXpathSingleFilter">
      <property name="checks" value="LineLength"/>
      <property name="query" value="//TEXT_BLOCK_LITERAL_BEGIN/.."/>
  </module>

com.puppycrawl.tools.checkstyle.api.CheckstyleException: SuppressionXpathSingleFilter is not allowed
```
I'm not sure if SuppressionXpathSingleFilter is the right filter to use in this case ? I saw this majorly being used in the other parts of the config file and so I was following them?

**Comment by Zopsss** [comment]
Solution provided by Nick at https://github.com/checkstyle/checkstyle/issues/11867#issuecomment-1180363780  is more easy and reliable

**Comment by mohitsatr** [comment]
@Zopsss yes, I tried the following suppression and it works 
```
   <module name="SuppressWithPlainTextCommentFilter">
        <property name="offCommentFormat" value='^.*"""((?!;).)*$'/>
        <property name="onCommentFormat" value='^\s+.*""";'/>
    </module>
```
but maybe the suppression suppresses too much and the following test fails
https://github.com/checkstyle/checkstyle/blob/c44cd647aec38213592573a598fc0b4e16f132d0/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4841indentation/IndentationTest.java#L113-L115

more specifically

https://github.com/checkstyle/checkstyle/blob/c44cd647aec38213592573a598fc0b4e16f132d0/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputCatchParametersOnNewLine.java#L55-L65

all the 3 violations above get suppressed.

**Comment by mohitsatr** [comment]
so I avoided the above suppression. `SuppressionXpathSingleFilter` seems more readable and widely used in the config.

**Comment by nrmancuso** [comment]
> [@nrmancuso](https://github.com/nrmancuso) [@romani](https://github.com/romani) [@strkkk](https://github.com/strkkk) should we add an example showing how to suppress multi-line string for LineLength check? I think it'll be beneficial for many users.

Yes, good idea

**Comment by Zopsss** [comment]
@mohitsatr try my above mentioned approach and also update documentation to add example for it.

**Comment by mohitsatr** [comment]
@Zopsss I tried your approach, no success. 

![Image](https://github.com/user-attachments/assets/fff4f0e9-f607-4279-9c61-a91378c7bdd4)

the reason it is not able detect OnComment is because OffComment's pattern match both opening and closing quotes in the following case:
https://github.com/checkstyle/checkstyle/blob/c44cd647aec38213592573a598fc0b4e16f132d0/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputCatchParametersOnNewLine.java#L55-L60

and both of the following if conditions are true and hence OnComment gets overwritten.

https://github.com/checkstyle/checkstyle/blob/87b7306edb88081276dbab80d9afed7819186172/src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilter.java#L277-L284

**Comment by mohitsatr** [comment]
I still think suppressing through XPath query is better. TEXT_BLOCK_LITERAL_BEGIN/TEXT_BLOCK_END looks more readable than regexp patterns. what do you think?

**Comment by Zopsss** [comment]
Please show the AST, let's try XPath approach

**Comment by mohitsatr** [comment]
```
    final String simpleScript = """
        s
        """;

|--VARIABLE_DEF -> VARIABLE_DEF [6:4]
        |       |   |--MODIFIERS -> MODIFIERS [6:4]
        |       |   |   `--FINAL -> final [6:4]
        |       |   |--TYPE -> TYPE [6:10]
        |       |   |   `--IDENT -> String [6:10]
        |       |   |--IDENT -> simpleScript [6:17]
        |       |   `--ASSIGN -> = [6:30]
        |       |       `--EXPR -> EXPR [6:32]
        |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [6:32]
        |       |               |--TEXT_BLOCK_CONTENT -> \n        s\n         [6:35]
        |       |               `--TEXT_BLOCK_LITERAL_END -> """ [8:8]
        |       |--SEMI -> ; [8:11]
```
```
    final String simpleScript1 = 
        """
        this is simple test;
        """;

        |       |--VARIABLE_DEF -> VARIABLE_DEF [10:4]
        |       |   |--MODIFIERS -> MODIFIERS [10:4]
        |       |   |   `--FINAL -> final [10:4]
        |       |   |--TYPE -> TYPE [10:10]
        |       |   |   `--IDENT -> String [10:10]
        |       |   |--IDENT -> simpleScript1 [10:17]
        |       |   `--ASSIGN -> = [10:31]
        |       |       `--EXPR -> EXPR [11:8]
        |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [11:8]
        |       |               |--TEXT_BLOCK_CONTENT -> \n        this is simple test;\n         [11:11]
        |       |               `--TEXT_BLOCK_LITERAL_END -> """ [13:8]
        |       |--SEMI -> ; [13:11]
```
```
    final String simpleScript2 = 
        """
        this is sample text """; 

        |       |--VARIABLE_DEF -> VARIABLE_DEF [15:4]
        |       |   |--MODIFIERS -> MODIFIERS [15:4]
        |       |   |   `--FINAL -> final [15:4]
        |       |   |--TYPE -> TYPE [15:10]
        |       |   |   `--IDENT -> String [15:10]
        |       |   |--IDENT -> simpleScript2 [15:17]
        |       |   `--ASSIGN -> = [15:31]
        |       |       `--EXPR -> EXPR [16:8]
        |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [16:8]
        |       |               |--TEXT_BLOCK_CONTENT -> \n        this is sample text  [16:11]
        |       |               `--TEXT_BLOCK_LITERAL_END -> """ [17:28]
        |       |--SEMI -> ; [17:31]
```
```
    return """
        this is sample text 
        """ 
            + simpleScript;

        |       |--LITERAL_RETURN -> return [19:4]
        |       |   |--EXPR -> EXPR [22:12]
        |       |   |   `--PLUS -> + [22:12]
        |       |   |       |--TEXT_BLOCK_LITERAL_BEGIN -> """ [19:11]
        |       |   |       |   |--TEXT_BLOCK_CONTENT -> \n        this is sample text \n         [19:14]
        |       |   |       |   `--TEXT_BLOCK_LITERAL_END -> """ [21:8]
        |       |   |       `--IDENT -> simpleScript [22:14]
        |       |   `--SEMI -> ; [22:26]
```

**Comment by romani** [comment]
@mohitsatr , I lost a clear definition of problem in this issue.
Please make comment with issue template to show behavior that author or issue is complaining about, I  will copy it to issue descripton.

You can create new  issue with more clear definition of problem.

**Comment by Zopsss** [comment]
@mohitsatr please try to create the xpath query, if you failed then ping me, I'll try to help you out.

**Comment by mohitsatr** [comment]
@romani please update the description 

the Original author wants LineLength Check to be disable for the conent of text blocks (""" ... """) 


> https://google.github.io/styleguide/javaguide.html#s4.8.9-text-blocks


![Image](https://github.com/user-attachments/assets/50d549bb-69a6-4b21-8bc5-42f693c74f5e)

```
public class LineLength {

  /** somejavadoc. */
  public static void main(String[] args) {
    
    // SHOULD NOT VIOLATE, THIS IS TEXT-BLOCK
    String textBlock = """ 
      This STRing will exceed the column limit of 100 but checktyle will not show many error because it is allowed 
      Random String
      google config allows text block to exceed limit of 100 line length. it improves readability and stuff.
      same is true for this line also. same is true for this line as well. same is true for this line too. 
      one more Random String.
        """;
     
    // OK to give violation as this is normal string
    String normalString = "a very long string is heavy to parse and takes so much processing power. it is not ideal to make long strings";
  }
}

$ java -jar checkstyle-10.25.0-all.jar -c google_checks.xml LineLength.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:1:1: Missing a Javadoc comment. [MissingJavadocType]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:7: Line is longer than 100 characters (found 115). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:9: Line is longer than 100 characters (found 108). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:10: Line is longer than 100 characters (found 107). [LineLength]
[WARN] /mnt/5D92528E6B945467/test/testing/LineLength.java:14: Line is longer than 100 characters (found 138). [LineLength]
Audit done.
```

you might have confused it with similar issue related to position of opening quotation of text-block https://github.com/google/google-java-format/issues/1258

**Comment by romani** [comment]
@mohitsatr , is this a workaround for problem ?
https://github.com/airlift/airbase/blob/2d4e1a72ac2707607317ca761c514d54919eedac/airbase-policy/src/main/resources/checkstyle/airbase-checks.xml#L6-L9


```
<module name="Checker">
    <module name="SuppressWithPlainTextCommentFilter">
        <property name="offCommentFormat" value='"""'/>
        <property name="onCommentFormat" value='"""'/>
    </module>
....
```

**Comment by mohitsatr** [comment]
> is this a workaround for problem

yes and no. I tried similar pattern using regexp, and it works but misses one edge case https://github.com/checkstyle/checkstyle/issues/11867#issuecomment-2924512013 . reason: https://github.com/checkstyle/checkstyle/issues/11867#issuecomment-2929016441

**Comment by mohitsatr** [comment]
I also tried `SuppressionXpathSingleFilter` but got exception because It requires the Check it is suppressing to be children of TreeWalker. LineLength Check is child of Checker. I don't know which suppression Filter is right one to use here ?
@romani

**Comment by Zopsss** [comment]
@mohitsatr try this:

https://checkstyle.org/filters/suppressionxpathfilter.html

**Comment by mohitsatr** [comment]
@Zopsss SuppressXpathFilter requires us to use an external suppression file. I think this Filter is for the users who have to have their own set of suppressions based on styling of their projects. 

One more thing:
> SuppressionXpathFilter can suppress Checks that have Treewalker as parent module.

I'm not this filter will work with checks having Checker as parent module

**Comment by Zopsss** [comment]
>> SuppressionXpathFilter can suppress Checks that have Treewalker as parent module.
>
> I'm not this filter will work with checks having Checker as parent module

Ah, I missed it. Thanks for pointing this out

**Comment by romani** [comment]
Try https://checkstyle.org/filters/suppresswithnearbytextfilter.html#Parent_Module

**Comment by mohitsatr** [comment]
@romani any filter which uses plain text or regex to suppress might not work in this problem. 
for google-style, we now know that opening and closing quotes must be on new lines, and if they are not, after #17159  resolves user will get violations from checkstyle. In this case we can easily use regex to make sure quotes are not preceded and followed by any thing i.e they are alone on the line.

but for users who don't use google-style, this solution will fail as they do not have the obligation to put quotes of text-blocks on new lines like the following case.

https://github.com/checkstyle/checkstyle/blob/c44cd647aec38213592573a598fc0b4e16f132d0/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputCatchParametersOnNewLine.java#L55-L65

this task is more suited for Filters with xpath.

**Comment by mohitsatr** [comment]
is extending LineLengthCheck a good idea ?

**Comment by Zopsss** [comment]
> we now know that opening and closing quotes must be on new lines

@mohitsatr FYI, the closing quotes doesn't always needs to be alone. It can be followed by some other code but it cannot be preceded with any character. 

For example:

```java
String s = 
            """
              Hello there! My name is:
            """ + getName();
```

https://google.github.io/styleguide/javaguide.html#s4.8.9-text-blocks

> The closing `"""` is on a new line with the same indentation as the opening `"""`, and may be followed on the same line by further code.

**Comment by mohitsatr** [comment]
> FYI, the closing quotes doesn't always needs to be alone

Thanks I missed that.
